odbc/src/api/connect_attr_tests.rs

```diff
@@ -0,0 +1,254 @@
+                        SQLGetConnectAttrW(
+                            mongo_handle as *mut _,
+                            attribute,
+                            value_ptr,
+                            buffer_length,
+                            out_length,
+                        )
+                    );
+
+                    $(assert_eq!($expected_length, *out_length);)?
+                    $(assert_eq!($expected_value, $actual_value_modifier(value_ptr, *out_length as usize));)?
+                }
+            }
+        };
+    }
+
+    unsafe fn modify_string_attr(value_ptr: Pointer, out_length: usize) -> String {
+        input_wtext_to_string(value_ptr as *const _, out_length)
+    }
+
+    unsafe fn modify_numeric_attr(value_ptr: Pointer, _: usize) -> u32 {
+        *(value_ptr as *mut UInteger)
+    }
+
+    test_get_attr!(
+        current_catalog_default,
+        attribute = ConnectionAttribute::CurrentCatalog,
+        expected_sql_return = SqlReturn::NO_DATA,
+    );
+
+    test_get_attr!(
+        current_catalog,
+        attribute = ConnectionAttribute::CurrentCatalog,
+        expected_sql_return = SqlReturn::SUCCESS,
+        initial_attrs = Box::new(ConnectionAttributes {
+            current_catalog: Some("test".to_string()),
+            ..Default::default()
+        }),
+        buffer_length = 5,
+        expected_length = 4,
+        expected_value = "test".to_string(),
+        actual_value_modifier = modify_string_attr,
+    );
+
+    test_get_attr!(
+        connection_timeout_default,
+        attribute = ConnectionAttribute::ConnectionTimeout,
+        expected_sql_return = SqlReturn::SUCCESS,
+        expected_value = 0u32,
+        actual_value_modifier = modify_numeric_attr,
+    );
+
+    test_get_attr!(
+        connection_timeout,
+        attribute = ConnectionAttribute::ConnectionTimeout,
+        expected_sql_return = SqlReturn::SUCCESS,
+        initial_attrs = Box::new(ConnectionAttributes {
+            connection_timeout: Some(42),
+            ..Default::default()
+        }),
+        expected_value = 42u32,
+        actual_value_modifier = modify_numeric_attr,
+    );
+
+    test_get_attr!(
+        login_timeout_default,
+        attribute = ConnectionAttribute::LoginTimeout,
+        expected_sql_return = SqlReturn::SUCCESS,
+        expected_value = 0u32,
+        actual_value_modifier = modify_numeric_attr,
+    );
+
+    test_get_attr!(
+        login_timeout,
+        attribute = ConnectionAttribute::LoginTimeout,
+        expected_sql_return = SqlReturn::SUCCESS,
+        initial_attrs = Box::new(ConnectionAttributes {
+            login_timeout: Some(42),
+            ..Default::default()
+        }),
+        expected_value = 42u32,
+        actual_value_modifier = modify_numeric_attr,
+    );
+
+    // Test getting unsupported attributes.
+    #[test]
+    fn get_unsupported_attr() {
+        unsafe {
+            for attr in [
+                ConnectionAttribute::AsyncEnable,
```

Comment:
The lists are different since `Set`'s list contains `CurrentCatalog` and `ConnectionTimeout`.

I can make the rest of the list a constant except for those two, and then for the `Set` test I can append the extra ones, so this will remove the duplication between parts of the two lists but waring that it won't be a wholesale replacement!

Reply:
done